File: code/14/solution.go
package main

import "fmt"

const upper = int64(1e7)

var val = [upper]int{}

func get(a int64) int {
	if a < upper && val[a] > 0 {
		return val[a]
	}
	if a == 1 {
		return 1
	}
	b := a >> 1
	if (a & 1) > 0 {
		b = a*3 + 1
	}

	v := get(b) + 1
	if a < upper {
		val[a] = v
	}
	return v
}

func main() {
	ans, ind := 0, 0
	for i := 1; i < 1000000; i++ {
		v := get(int64(i))
		if v > ans {
			ans = v
			ind = i
		}
	}
	fmt.Printf("%d\n", ind)
}

File: code/14/solution.cpp
#include <iostream>
#include <map>
#include <set>
#include <vector>
using namespace std;
const int upper = 1e7;
int val[upper];
int get(long long a) {
    if (a < upper && val[a])
        return val[a];
    if (a == 1)
        return 1;
    long long b = a >> 1;
    if (a & 1)
        b = a * 3 + 1;

    int v = get(b) + 1;
    if (a < upper)
        val[a] = v;
    return v;
}
int main() {
    int ans = 0;
    int ind = 0;
    for (int i = 1; i < 1000000; i++) {
        int v = get(i);
        if (v > ans) {
            ans = v;
            ind = i;
        }
    }
    cout << ind << endl;
    return 0;
}
